common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.cpp

```diff
@@ -509,9 +552,146 @@ bool base_wlan_hal_dwpal::attach_ctrl_interface(int vap_id)
+                                 (enum ltq_nl80211_vendor_subcmds)nl_cmd, vendor_data,
+                                 vendor_data_size) != DWPAL_SUCCESS) {
+        LOG(ERROR) << __FUNCTION__ << "ERROR for cmd = " << nl_cmd;
+        return false;
+    }
+
+    return true;
+}
+
+bool base_wlan_hal_dwpal::dwpal_nl_cmd_get(const std::string &ifname, unsigned int nl_cmd,
+                                           unsigned char *out_buffer, size_t *out_buffer_size,
+                                           const size_t max_buffer_size)
+{
+
+    if ((out_buffer_size == NULL) || (out_buffer == NULL)) {
+        LOG(ERROR) << __FUNCTION__ << "out_buffer or out_buffer_size incorrect ==> Abort!\n";
+        return false;
+    }
+
+    /* Handle a command which invokes an event with the output data */
+    if (dwpal_driver_nl_cmd_send(
+            m_dwpal_nl_ctx, DWPAL_NL_SOLICITED_EVENT, (char *)ifname.c_str(), NL80211_CMD_VENDOR,
+            DWPAL_NETDEV_ID, (enum ltq_nl80211_vendor_subcmds)nl_cmd, NULL, 0) == DWPAL_FAILURE) {
+        LOG(ERROR) << __FUNCTION__ << "ERROR for cmd = " << nl_cmd;
+        return false;
+    }
+
+    // Passing a lambda with capture is not supported for standard C function
+    // pointers. As a workaround, we create a static (but thread local) wrapper
+    // function that calls the capturing lambda function.
+    static __thread std::function<DWPAL_Ret(char *ifname, int event, int subevent, size_t len,
+                                            unsigned char *data)>
+        nl_handler_cb_wrapper;
+    nl_handler_cb_wrapper = [&](char *ifname, int event, int subevent, size_t len,
+                                unsigned char *data) -> DWPAL_Ret {
+        LOG(DEBUG) << __func__ << " ifname = " << std::string(ifname) << " event = " << event
+                   << " subevent = " << subevent << " len = " << len;
+
+        if ((len == 0) || (data == NULL)) {
+            LOG(ERROR) << __func__ << "len=0 and/or data is NULL ==> Abort!\n";
+            return DWPAL_FAILURE;
+        }
+        if (event == NL80211_CMD_VENDOR) {
+            if (len >= max_buffer_size) {
+                LOG(ERROR) << __func__ << "NL size exceeds out_buffer size ==> Abort!\n";
+                return DWPAL_FAILURE;
+            }
+            // copy get result from nl buffer, and sets the length copied to out_buffer_size
+            *out_buffer_size = std::copy_n(data,len,out_buffer) - out_buffer;
+        } else {
+            LOG(ERROR) << __func__ << "non vendor event not handled ";
+            return DWPAL_FAILURE;
+        }
+        return DWPAL_SUCCESS;
+    };
+    auto nl_handler_cb = [](char *ifname, int event, int subevent, size_t len,
+                            unsigned char *data) -> DWPAL_Ret {
+        return nl_handler_cb_wrapper(ifname, event, subevent, len, data);
+    };
+
+    while (*out_buffer_size == 0) { /* wait for reply with valid data */
```

Comment:
I don't understand this hack. Why not just call the API and be done with it?

Reply:
Because you're waiting for a specific message, and should continue parsing messages until you get it.
On the other hand, we shouldn't wait indefinitely. Some sort of timeout mechanism should be implemented.

Reply:
based on testing in RDKB platform.
removed for now (at least until we test  again with single code base inRDKB.)